File: frontend/src/app/core/sim/data/ship-catalog.ts
import { ShipTypeDef } from '../../models';

/**
 * Kontersystem, siehe Mechanik/03_..., §2 (verbindlich, überschreibt die
 * frühere Formulierung in Konzeption/03_...): Korvette schlägt Kreuzer,
 * Kreuzer schlägt Zerstörer, Zerstörer schlägt Korvette. Der jeweilige
 * Multiplikator (×2 im Vorteil, ×0,5 im Nachteil, sonst ×1) wird erst im
 * Kampfsystem angewendet (Mechanik/04_..., §4) — hier steht nur, wer wen
 * kontert. Keine attack/hull-Werte: militärischer Wert = Produktionsaufwand,
 * siehe ShipTypeDef.
 */
export const SHIP_CATALOG: ShipTypeDef[] = [
  { productTypeId: 'p_corvette', class: 'Corvette', cargoMassKg: 0, cargoVolumeM3: 0, carrierSlotUsage: 1, countersClass: 'Cruiser' },
  { productTypeId: 'p_destroyer', class: 'Destroyer', cargoMassKg: 0, cargoVolumeM3: 0, carrierSlotUsage: 2, countersClass: 'Corvette' },
  { productTypeId: 'p_cruiser', class: 'Cruiser', cargoMassKg: 0, cargoVolumeM3: 0, carrierSlotUsage: 4, countersClass: 'Destroyer' },
  // Grobe Erstschätzung, siehe ProductType-Massen/Volumina in product-catalog.ts
  // (z. B. 1 Einheit Ferrometallerz = 800.000 kg – "genaue Zahl offen").
  { productTypeId: 'p_freighter', class: 'Freighter', cargoMassKg: 2_000_000, cargoVolumeM3: 3000, carrierSlotUsage: 3, countersClass: null },
  { productTypeId: 'p_carrier', class: 'Carrier', cargoMassKg: 0, cargoVolumeM3: 0, carrierSlotUsage: 0, countersClass: null },
  { productTypeId: 'p_trooptransport', class: 'TroopTransport', cargoMassKg: 0, cargoVolumeM3: 0, carrierSlotUsage: 2, countersClass: null },
];

export function findShipDef(productTypeId: string): ShipTypeDef {
  const found = SHIP_CATALOG.find(s => s.productTypeId === productTypeId);
  if (!found) throw new Error(`Unbekannter ShipType: ${productTypeId}`);
  return found;
}

File: backend/src/main/java/de/nebula/data/ShipCatalog.java
package de.nebula.data;

import com.fasterxml.jackson.core.type.TypeReference;
import de.nebula.engine.GameConstants;
import de.nebula.model.ShipTypeDef;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * Schiffskatalog – die Daten liegen in {@code shared/catalog/ships.json}
 * (siehe {@link CatalogJson}). Kontersystem, siehe Mechanik/03_..., §2
 * (verbindlich): Korvette schlägt Kreuzer, Kreuzer schlägt Zerstörer,
 * Zerstörer schlägt Korvette – abgebildet über das Feld {@code countersClass}
 * in der JSON-Datei. Der Multiplikator (×2 im Vorteil, ×0,5 im Nachteil,
 * sonst ×1) wird erst im Kampfsystem angewendet (Mechanik/04_..., §4).
 * Keine attack/hull-Werte: militärischer Wert = Produktionsaufwand.
 */
public final class ShipCatalog {
  private ShipCatalog() {
  }

  public static final List<ShipTypeDef> CATALOG = List.copyOf(withJumpFuel(
      CatalogJson.load("ships.json", new TypeReference<List<ShipTypeDef>>() {
      })));

  /**
   * Rechnet Sprungverbrauch und Tankgröße JE SCHIFF aus seiner Masse aus
   * (Umsetzungskonzept/34_...md, Entscheidung F7). Bewusst hier beim Laden und
   * nicht als weitere Spalte in {@code ships.json}: die Masse steht im
   * Produktkatalog, und zwei gepflegte Zahlen für dieselbe Aussage laufen
   * auseinander, sobald ein Schiff neu vermessen wird.
   */
  private static List<ShipTypeDef> withJumpFuel(List<ShipTypeDef> defs) {
    double corvetteMass = ProductCatalog.find(GameConstants.CORVETTE_PRODUCT_ID).massKg;
    for (ShipTypeDef def : defs) {
      double massRatio = corvetteMass > 0 ? ProductCatalog.find(def.productTypeId).massKg / corvetteMass : 1;
      def.jumpFuelPerHop = massRatio * GameConstants.JUMP_FUEL_PER_CORVETTE_MASS_PER_HOP;
      def.fuelTankCapacity = def.jumpFuelPerHop * GameConstants.JUMP_FUEL_TANK_RANGE_HOPS;
    }
    return defs;
  }

  /** Index nach Produkt-Id – siehe {@code ProductCatalog.BY_ID}: {@link #find} läuft in jedem Tick für jede Schiffsgruppe. */
  private static final Map<String, ShipTypeDef> BY_ID = CATALOG.stream()
      .collect(Collectors.toUnmodifiableMap(s -> s.productTypeId, s -> s));

  public static ShipTypeDef find(String productTypeId) {
    ShipTypeDef def = BY_ID.get(productTypeId);
    if (def == null) throw new IllegalArgumentException("Unbekannter ShipType: " + productTypeId);
    return def;
  }
}
